File: canonical-collaboration-demo/scripts/dev.py
from __future__ import annotations

import os
from pathlib import Path
import signal
import subprocess
import sys
import time


ROOT = Path(__file__).resolve().parent.parent


def load_env() -> dict[str, str]:
    env = os.environ.copy()
    env_file = ROOT / ".env"
    if not env_file.exists():
        print("Missing .env. Run: cp .env.example .env", file=sys.stderr)
        raise SystemExit(1)
    for line in env_file.read_text().splitlines():
        line = line.strip()
        if not line or line.startswith("#") or "=" not in line:
            continue
        key, value = line.split("=", 1)
        env.setdefault(key.strip(), value.strip().strip('"').strip("'"))
    if not env.get("OPENAI_API_KEY"):
        print("Add OPENAI_API_KEY to .env before starting the demo.", file=sys.stderr)
        raise SystemExit(1)
    collaboration_port = env.get("COLLABORATION_PORT", "18081")
    env.update(
        {
            "PORT": collaboration_port,
            "COLLABORATION_URL": f"ws://127.0.0.1:{collaboration_port}",
            "PUBLIC_COLLABORATION_URL": f"ws://localhost:{collaboration_port}",
            "DOCUMENT_ROOT": str(ROOT / "data"),
            "VITE_API_URL": "http://localhost:8000",
        }
    )
    return env


def main() -> None:
    env = load_env()
    commands = [
        ("collaboration", ["npm", "start"], ROOT / "collaboration"),
        ("api", [str(ROOT / ".venv/bin/uvicorn"), "app.main:app", "--host", "0.0.0.0", "--port", "8000"], ROOT / "server"),
        ("client", ["npm", "run", "dev"], ROOT / "client"),
    ]
    processes: list[tuple[str, subprocess.Popen[bytes]]] = []
    stopping = False

    def stop(*_: object) -> None:
        for _, process in processes:
            if process.poll() is None:
                process.terminate()
        deadline = time.monotonic() + 5
        for _, process in processes:
            try:
                process.wait(timeout=max(0.1, deadline - time.monotonic()))
            except subprocess.TimeoutExpired:
                process.kill()

    def request_stop(*_: object) -> None:
        nonlocal stopping
        stopping = True

    signal.signal(signal.SIGINT, request_stop)
    signal.signal(signal.SIGTERM, request_stop)

    try:
        for name, command, cwd in commands:
            process = subprocess.Popen(command, cwd=cwd, env=env)
            processes.append((name, process))
            print(f"[dev] started {name} (pid {process.pid})", flush=True)
        print("[dev] open http://localhost:15173", flush=True)

        while not stopping:
            for name, process in processes:
                return_code = process.poll()
                if return_code is not None:
                    print(f"[dev] {name} exited with status {return_code}", file=sys.stderr)
                    raise SystemExit(return_code or 1)
            time.sleep(0.25)
    finally:
        stop()


if __name__ == "__main__":
    main()

File: canonical-collaboration-demo/scripts/dev.mjs
import fs from 'node:fs';
import path from 'node:path';
import { spawn } from 'node:child_process';
import { fileURLToPath } from 'node:url';

const root = path.resolve(path.dirname(fileURLToPath(import.meta.url)), '..');
const validLogServices = new Set(['all', 'api', 'rooms', 'worker', 'agent', 'documentworker', 'collab', 'client', 'dev']);

function parseLogServices(arguments_) {
  const logIndex = arguments_.indexOf('--log');
  if (logIndex === -1) return new Set(['all']);
  const value = arguments_[logIndex + 1];
  if (!value || value.startsWith('--')) throw new Error('--log requires a comma-separated service list.');
  const services = new Set(value.split(',').map((service) => service.trim().replaceAll('-', '')).filter(Boolean));
  const invalid = [...services].filter((service) => !validLogServices.has(service));
  if (invalid.length) {
    throw new Error(`Unknown log service: ${invalid.join(', ')}. Choose from: ${[...validLogServices].join(', ')}.`);
  }
  return services;
}

const logServices = parseLogServices(process.argv.slice(2));
const logs = (service) => logServices.has('all') || logServices.has(service);

function loadEnvironment() {
  const environment = { ...process.env };
  const envPath = path.join(root, '.env');
  if (!fs.existsSync(envPath)) throw new Error('Missing .env. Run: cp .env.example .env');
  for (const sourceLine of fs.readFileSync(envPath, 'utf8').split(/\r?\n/)) {
    const line = sourceLine.trim();
    if (!line || line.startsWith('#') || !line.includes('=')) continue;
    const separator = line.indexOf('=');
    const key = line.slice(0, separator).trim();
    const value = line.slice(separator + 1).trim().replace(/^(['"])(.*)\1$/, '$2');
    environment[key] ??= value;
  }
  if (!environment.OPENAI_API_KEY) throw new Error('Add OPENAI_API_KEY to .env before starting the demo.');
  return {
    ...environment,
    PORT: '8000',
    PUBLIC_ORIGIN: 'http://localhost:8000',
    DOCUMENT_ROOT: path.join(root, 'data'),
    VITE_API_URL: 'http://localhost:8000',
    LOG_SERVICES: [...logServices].join(','),
  };
}

const environment = loadEnvironment();
const services = [
  {
    name: 'server',
    command: process.execPath,
    args: ['src/server.js'],
    cwd: path.join(root, 'server'),
    showOutput: [...logServices].some((service) => service !== 'client' && service !== 'dev') || logServices.has('all'),
  },
  { name: 'client', command: 'npm', args: ['run', 'dev'], cwd: path.join(root, 'client'), showOutput: logs('client') },
];
const processes = [];
let stopping = false;

function stop(exitCode = 0) {
  if (stopping) return;
  stopping = true;
  for (const child of processes) child.kill('SIGTERM');
  setTimeout(() => {
    for (const child of processes) if (child.exitCode === null) child.kill('SIGKILL');
  }, 5_000).unref();
  if (exitCode) process.exitCode = exitCode;
}

for (const service of services) {
  const child = spawn(service.command, service.args, {
    cwd: service.cwd,
    env: environment,
    stdio: ['inherit', service.showOutput ? 'inherit' : 'ignore', 'inherit'],
  });
  processes.push(child);
  if (logs('dev')) console.log(`[dev] started ${service.name} (pid ${child.pid})`);
  child.once('exit', (code, signal) => {
    if (!stopping) {
      console.error(`[dev] ${service.name} exited (${signal ?? code})`);
      stop(code || 1);
    }
  });
}

if (logs('dev')) console.log('[dev] open http://localhost:15173');
process.once('SIGINT', () => stop());
process.once('SIGTERM', () => stop());
